src/main/java/de/numcodex/sq2cql/model/structured_query/AbstractCriterion.java

```diff
@@ -28,133 +29,174 @@
+   */
+  static Container<RetrieveExpression> retrieveExpr(MappingContext mappingContext,
+      TermCode termCode) {
+    return codeSelector(mappingContext, termCode).map(terminology -> {
+      var mapping = mappingContext.findMapping(termCode)
+          .orElseThrow(() -> new MappingNotFoundException(termCode));
+      return RetrieveExpression.of(mapping.resourceType(), terminology);
+    });
+  }
+
+
+  static Container<RetrieveExpression> retrieveResourceTypeExpr(MappingContext mappingContext,
+      TermCode termCode) {
+    return codeSelector(mappingContext, termCode).map(terminology -> {
+      var mapping = mappingContext.findMapping(termCode)
+          .orElseThrow(() -> new MappingNotFoundException(termCode));
+      return RetrieveExpression.of(mapping.resourceType());
+    });
+  }
+
+
+  static Container<BooleanExpression> modifiersExpr(List<Modifier> modifiers,
+      MappingContext mappingContext,
+      IdentifierExpression identifier) {
+    return modifiers.stream()
+        .map(m -> m.expression(mappingContext, identifier))
+        .reduce(Container.empty(), Container.AND);
+  }
+
+  static ExistsExpression existsExpr(SourceClause sourceClause, BooleanExpression whereExpr) {
+    return ExistsExpression.of(QueryExpression.of(sourceClause, WhereClause.of(whereExpr)));
+  }
+
+  public Concept getConcept() {
+    return concept;
+  }
+
+  @Override
+  public Container<BooleanExpression> toCql(MappingContext mappingContext) {
+    var expr = fullExpr(mappingContext);
+    if (expr.isEmpty()) {
+      throw new TranslationException("Failed to expand the concept %s.".formatted(concept));
     }
-
-    private Container<BooleanExpression> expr(MappingContext mappingContext, TermCode termCode) {
-        var mapping = mappingContext.findMapping(termCode).orElseThrow(() -> new MappingNotFoundException(termCode));
-        if ("Patient".equals(mapping.resourceType())) {
-            return valueAndModifierExpr(mappingContext, mapping, PATIENT);
+    return expr;
+  }
+
+  /**
+   * Builds an OR-expression with an expression for each concept of the expansion of {@code
+   * termCode}.
+   */
+  private Container<BooleanExpression> fullExpr(MappingContext mappingContext) {
+    return mappingContext.expandConcept(concept)
+        .map(termCode -> expr(mappingContext, termCode))
+        .reduce(Container.empty(), Container.OR);
+  }
+
+  private Container<BooleanExpression> expr(MappingContext mappingContext, TermCode termCode) {
+    var mapping = mappingContext.findMapping(termCode)
+        .orElseThrow(() -> new MappingNotFoundException(termCode));
+    if ("Patient".equals(mapping.resourceType())) {
+      return valueAndModifierExpr(mappingContext, mapping, PATIENT);
+    } else if ("MedicationAdministration".equals(mapping.resourceType())) {
+      return retrieveResourceTypeExpr(mappingContext, termCode).flatMap(retrieveExpr -> {
+        var self = ReferenceCriterion.from(this, termCode);
```

Comment:
I see that you use some code from AbstractCriterion here. But at the end, there is no ReferenceCriterion. Please put the code generating the expression container into a class which isn't a criterion.